joblib/test/test_memory.py

```diff
@@ -417,38 +405,49 @@ def test_persistence(tmpdir):
     gp(1)
 
 
-def test_call_and_shelve(tmpdir):
+@fixture(scope='function', params=range(4))
+def testcase_call_and_shelve(request, tmpdir):
+    func = [MemorizedFunc(f, tmpdir.strpath),
+            NotMemorizedFunc(f),
+            Memory(cachedir=tmpdir.strpath).cache(f),
+            Memory(cachedir=None).cache(f)]
+
+    Result = [MemorizedResult, NotMemorizedResult,
+              MemorizedResult, NotMemorizedResult]
+
+    return func[request.param], Result[request.param]
+
+
+def test_call_and_shelve(testcase_call_and_shelve):
     """Test MemorizedFunc outputting a reference to cache.
     """
+    func, Result = testcase_call_and_shelve
+    assert func(2) == 5
+    result = func.call_and_shelve(2)
+    assert isinstance(result, Result)
+    assert result.get() == 5
 
-    for func, Result in zip((MemorizedFunc(f, tmpdir.strpath),
-                             NotMemorizedFunc(f),
-                             Memory(cachedir=tmpdir.strpath).cache(f),
-                             Memory(cachedir=None).cache(f),
-                             ),
-                            (MemorizedResult, NotMemorizedResult,
-                             MemorizedResult, NotMemorizedResult)):
-        assert func(2) == 5
-        result = func.call_and_shelve(2)
-        assert isinstance(result, Result)
-        assert result.get() == 5
-
-        result.clear()
-        with raises(KeyError):
-            result.get()
-        result.clear()  # Do nothing if there is no cache.
-
-
-def test_memorized_pickling(tmpdir):
-    for func in (MemorizedFunc(f, tmpdir.strpath), NotMemorizedFunc(f)):
-        filename = tmpdir.join('pickling_test.dat').strpath
-        result = func.call_and_shelve(2)
-        with open(filename, 'wb') as fp:
-            pickle.dump(result, fp)
-        with open(filename, 'rb') as fp:
-            result2 = pickle.load(fp)
-        assert result2.get() == result.get()
-        os.remove(filename)
+    result.clear()
+    with raises(KeyError):
+        result.get()
+    result.clear()  # Do nothing if there is no cache.
+
+
+@fixture(scope='function', params=range(2))
```

Comment:
Okay I'll keep for loops until there's something interesting other than indirect parametrization from pytest's side.